src/io.h

```diff
@@ -211,63 +204,33 @@ class FormatXTC {
      * @note The container particle vector *must* match the number of particles
      *       in the xtc file. If not
      *       an error message will be issued and the function will abort.
-     *       You may want to transfer the new box size to the pair potential if
-     *       periodic boundaries are used.
      */
-    template <class Tspace> bool loadNextFrame(Tspace &c, bool setbox = true, bool applypbc = false) {
-        if (xd != nullptr) {
-            if (natoms_xtc == (int)c.p.size()) {
-                int rc = read_xtc(xd, natoms_xtc, &step_xtc, &time_xtc, xdbox, x_xtc, &prec_xtc);
-                if (rc == 0) {
-                    // Geometry::Chameleon *geo = dynamic_cast<Geometry::Chameleon *>(&c.geo);
-                    // if (geo == nullptr or geo->type not_eq Geometry::CUBOID)
-                    //    throw std::runtime_error("Cuboid-like geometry required");
-                    Point len_half = 0.5 * c.geo.getLength();
-                    if (setbox)
-                        c.geo.setLength(Point(10.0 * xdbox[0][0], 10.0 * xdbox[1][1], 10.0 * xdbox[2][2]));
-                    for (size_t i = 0; i < c.p.size(); i++) {
-                        c.p[i].pos.x() = 10.0 * x_xtc[i][0];
-                        c.p[i].pos.y() = 10.0 * x_xtc[i][1];
-                        c.p[i].pos.z() = 10.0 * x_xtc[i][2];
-                        c.p[i].pos -= len_half;
-                        if (applypbc)
-                            c.geo.boundary(c.p[i].pos);
-                        if (c.geo.collision(c.p[i].pos, 0))
-                            throw std::runtime_error("particle-container collision");
-                    }
-                    return true;
-                }
-            } else
-                throw std::runtime_error("xtcfile<->container particle mismatch");
-        } else
-            throw std::runtime_error("xtc file cannot be read");
-        return false; // end of file or not opened
-    }
+    bool loadNextFrame(Space &spc, bool setbox = true, bool apply_periodic_boundaries = false);
 
     /**
      * This will take an arbitrary particle vector and add it
      * to an xtc file. If the file is already open, coordinates will
      * be added, while a new file is created if not.
      * Coordinates are shifted and converted to nanometers.
      * Box dimensions for the frame must be manually
-     * set by the `setbox()` function before calling this.
+     * set by the `setLength()` function before calling this.
      */
-    template <class Titer1, class Titer2 /** particle vector iterator */>
-    bool save(const std::string &file, Titer1 begin, Titer2 end) {
+    template <class begin_iterator, class end_iterator /** particle vector iterator */>
+    bool save(const std::string &file, begin_iterator begin, end_iterator end) {
```

Comment:
I would change the responsibilities: The file parameter should pass to the constructor. Here only the iterator shall be the argument. And the method should be renamed to `write`. The we could write the trajectory incrementally during the run.

Reply:
Mentioned in todo in class header. The XTC handling is arcane and very outdated. It would need to be re-written in a separate PR.